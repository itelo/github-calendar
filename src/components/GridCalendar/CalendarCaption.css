.calendar-caption-container {
  width: 130px;
  display: flex;
  flex-direction: row;
  font-size: 11px;
  color: #586069;
  align-items: center;
  justify-content: space-between;
}

.calendar-caption-squares-container {
  display: flex;
  flex-direction: row;
  width: 62px;
  justify-content: space-between;
}
