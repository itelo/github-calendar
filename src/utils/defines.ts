export const CONTRIBUTIONS_MOST = 35;
export const CONTRIBUTIONS_ALOT = 28;
export const CONTRIBUTIONS_MODERATE = 21;
export const CONTRIBUTIONS_LITTLE = 14;

export const MAX_DAYS_IN_YEAR = 366;
export const MAX_DAYS_IN_WEEK = 7;

export const TOTAL_WEEK_DAYS = 7;
export const TOTAL_MONTH_COUNT = 12;
