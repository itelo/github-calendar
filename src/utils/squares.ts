import {
  CONTRIBUTIONS_MOST,
  CONTRIBUTIONS_ALOT,
  CONTRIBUTIONS_MODERATE,
  CONTRIBUTIONS_LITTLE
} from "./defines";

export const getColorClass = (count: number) => {
  switch (true) {
    case count >= CONTRIBUTIONS_MOST:
      return "most";
    case count >= CONTRIBUTIONS_ALOT:
      return "alot";
    case count >= CONTRIBUTIONS_MODERATE:
      return "moderate";
    case count >= CONTRIBUTIONS_LITTLE:
      return "little";
    default:
      return "nothing";
  }
};
