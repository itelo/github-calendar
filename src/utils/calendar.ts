import dfns from "date-fns";
import * as R from "ramda";
import { MAX_DAYS_IN_YEAR } from "./defines";

export const getAllDaysBasedInAYear = (year: number) => {
  const today = new Date();
  const yearHelp = dfns.setYear(today, year);
  const maxDateBasedOnYearPassed = dfns.endOfYear(yearHelp);
  const lastDay = dfns.isAfter(maxDateBasedOnYearPassed, today)
    ? today
    : maxDateBasedOnYearPassed;

  const possibleFirstDay = dfns.subDays(lastDay, MAX_DAYS_IN_YEAR);

  const firstDay = findSaturday(possibleFirstDay);

  const rangeDays = dfns.differenceInCalendarDays(lastDay, firstDay);
  const rangeDaysToSubtract = R.reverse(R.range(0, rangeDays));

  return R.map(subtract(lastDay), rangeDaysToSubtract);
};

export const findSaturday = (date: Date) => {
  let count = 0;
  for (;;) {
    if (!dfns.isSaturday(dfns.subDays(date, count))) {
      count++;
    } else {
      break;
    }
  }
  return dfns.subDays(date, count);
};

export const subtract = (date: Date) => (day: number) =>
  dfns.subDays(date, day);
type TreeDays = {
  date: string;
  count: number;
};

export const concat = (treeDays: TreeDays[]) => (day: Date) => {
  const data = R.find(isDayInTreeDays(day), treeDays);
  if (data) {
    return {
      date: day,
      count: data.count
    };
  } else {
    return {
      date: day,
      count: 0
    };
  }
};

export const isDayInTreeDays = (day: Date) => ({ date }: { date: string }) =>
  dfns.isSameDay(new Date(date), day);

export const mergeTreeJsonWithAllDays = (
  allDays: Date[],
  treeDays: TreeDays[]
) => R.map(concat(treeDays), allDays);

export const extractMonthPositionFromData = (
  acc: { month: string; position: number; lastMonthHelper: string }[],
  data: {
    count: number;
    date: Date;
  }[],
  index: number
) => {
  const month = dfns.format(data[0].date, "MMM");

  if (acc.length === 0 || acc[acc.length - 1].lastMonthHelper !== month) {
    acc.push({
      month,
      position: index,
      lastMonthHelper: month
    });
  }
  return acc;
};
