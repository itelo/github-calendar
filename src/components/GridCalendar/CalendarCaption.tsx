import React from "react";
import {
  CONTRIBUTIONS_MOST,
  CONTRIBUTIONS_ALOT,
  CONTRIBUTIONS_MODERATE,
  CONTRIBUTIONS_LITTLE
} from "../../utils/defines";
import { getColorClass } from "../../utils/squares";
import "./CalendarCaption.css";

export const allPossibilities = [
  0,
  CONTRIBUTIONS_LITTLE,
  CONTRIBUTIONS_MODERATE,
  CONTRIBUTIONS_ALOT,
  CONTRIBUTIONS_MOST
];

const CalendarCaption = () => {
  return (
    <div
      data-testid="calendar-caption-container"
      className="calendar-caption-container"
    >
      Less
      <div className="calendar-caption-squares-container">
        {allPossibilities.map(possibility => (
          <div
            data-testid={`possibility-${possibility}`}
            key={possibility}
            className={`big-square ${getColorClass(possibility)}`}
          />
        ))}
      </div>
      More
    </div>
  );
};

export default CalendarCaption;
