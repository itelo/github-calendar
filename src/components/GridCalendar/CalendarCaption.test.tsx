import React from "react";
import { render, fireEvent } from "react-testing-library";
import "react-testing-library/cleanup-after-each";
import CalendarCaption, { allPossibilities } from "./CalendarCaption";
import "jest-dom/extend-expect";
import { getColorClass } from "../../utils/squares";

describe("<CalendarCaption />", () => {
  test("test calendar caption using allPossibilities and getColorClass", () => {
    const { getByTestId } = render(<CalendarCaption />);

    allPossibilities.forEach(possibility => {
      const content = getByTestId(`possibility-${possibility}`);
      expect(content).toHaveClass("big-square");
      expect(content).toHaveClass(getColorClass(possibility));
    });

    const container = getByTestId("calendar-caption-container");
    expect(container).toHaveTextContent("LessMore");
  });

  test("just snapshot", () => {
    const { container } = render(<CalendarCaption />);

    expect(container).toMatchSnapshot();
  });
});
