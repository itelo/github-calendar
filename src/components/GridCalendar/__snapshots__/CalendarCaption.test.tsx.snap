// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalendarCaption /> just snapshot 1`] = `
<div>
  <div
    class="calendar-caption-container"
    data-testid="calendar-caption-container"
  >
    Less
    <div
      class="calendar-caption-squares-container"
    >
      <div
        class="big-square nothing"
        data-testid="possibility-0"
      />
      <div
        class="big-square little"
        data-testid="possibility-14"
      />
      <div
        class="big-square moderate"
        data-testid="possibility-21"
      />
      <div
        class="big-square alot"
        data-testid="possibility-28"
      />
      <div
        class="big-square most"
        data-testid="possibility-35"
      />
    </div>
    More
  </div>
</div>
`;
